src/libraries/Microsoft.PowerFx.Interpreter/PatchFunction.cs

```diff
@@ -0,0 +1,182 @@
+                {
+                    fieldsDict[field.Name] = field.Value;
+                }
+            }
+
+            var fieldList = new List<NamedValue>();
+
+            foreach (var field in fieldsDict)
+            {
+                fieldList.Add(new NamedValue(field.Key, field.Value));
+            }
+
+            var record = RecordValue.NewRecordFromFields(fieldList);
+
+            return Task.FromResult<FormulaValue>(record);
+        }
+
+        public override RequiredDataSourcePermissions FunctionPermission => RequiredDataSourcePermissions.Create | RequiredDataSourcePermissions.Update;
+    }
+
+    // Patch( DataSource, BaseRecord, ChangeRecord1 [, ChangeRecord2, … ])
+    internal class PatchFunction : PatchAndValidateRecordFunctionBase, IAsyncTexlFunction
+    {
+        public override bool IsSelfContained => false;
+
+        public PatchFunction()
+            : base("Patch", TexlStrings.AboutPatch, FunctionCategories.Table | FunctionCategories.Behavior, DType.EmptyRecord, 0, 3, int.MaxValue, DType.EmptyTable, DType.EmptyRecord, DType.EmptyRecord)
+        {
+        }
+
+        public override IEnumerable<TexlStrings.StringGetter[]> GetSignatures()
+        {
+            yield return new[] { TexlStrings.PatchDataSourceArg, TexlStrings.PatchBaseRecordArg };
+            yield return new[] { TexlStrings.PatchDataSourceArg, TexlStrings.PatchBaseRecordArg, TexlStrings.PatchChangeRecordsArg };
+            yield return new[] { TexlStrings.PatchDataSourceArg, TexlStrings.PatchBaseRecordArg, TexlStrings.PatchChangeRecordsArg, TexlStrings.PatchChangeRecordsArg };
+        }
+
+        public override IEnumerable<TexlStrings.StringGetter[]> GetSignatures(int arity)
+        {
+            if (arity > 3)
+            {
+                return GetGenericSignatures(arity, TexlStrings.PatchDataSourceArg, TexlStrings.PatchBaseRecordArg, TexlStrings.PatchChangeRecordsArg);
+            }
+
+            return base.GetSignatures(arity);
+        }
+
+        public async Task<FormulaValue> InvokeAsync(FormulaValue[] args, CancellationToken cancel)
+        {
+            var argFields = new Dictionary<string, FormulaValue>();
+
+            var arg0 = args[0] as TableValue;
+            var arg1 = args[1] as RecordValue;
+
+            // Starts from the 3rd argument (new values)
+            for (var i = 2; i < args.Length; i++)
+            {
+                var arg = args[i];
+
+                if (arg is BlankValue)
+                {
+                    continue;
+                }
+                else if (arg is ErrorValue)
+                {
+                    // !JYL! How to handle error value arguments?
+                    return arg;
+                }
+                else if (arg is RecordValue record)
+                {
+                    foreach (var field in record.Fields)
+                    {
+                        argFields[field.Name] = field.Value;
+                    }
+                }
```

Comment:
>} [](http://example.com/codeflow?start=16&length=1)

We should add tests where these args are neither Blank, nor an ErrorValue nor a RecordValue

Reply:
Those should be binding errors. 
So we could just have an else here and throw an exception. 
(it's ok to throw exceptions when somebody circumvents the binder by calling SDK methods direclty)